core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java

```diff
@@ -177,192 +166,313 @@ public DefaultRequestLog(RequestContext ctx, ContentPreviewerFactory requestCont
-        lastChild.addListener(this::propagateResponseEndData, RESPONSE_END);
+        // Safe to allow using as CompletableFuture<RequestOnlyLog>
+        // because a user is never allowed to complete.
+        @SuppressWarnings("unchecked")
+        final CompletableFuture<T> cast = (CompletableFuture<T>) newFuture;
+        return cast;
     }
 
-    private void propagateResponseEndData(RequestLog log) {
-        responseLength(log.responseLength());
-        responseContentPreview(log.responseContentPreview());
-        responseTrailers(log.responseTrailers());
-        endResponse0(log.responseCause(), log.responseEndTimeNanos());
+    @Override
+    public CompletableFuture<RequestOnlyLog> requestCompleteFuture() {
+        return future(RequestLogProperty.FLAGS_REQUEST_COMPLETE);
     }
 
     @Override
-    public List<RequestLog> children() {
-        return children != null ? ImmutableList.copyOf(children) : ImmutableList.of();
+    public CompletableFuture<RequestLog> partialFuture(RequestLogProperty property) {
+        return future(requireNonNull(property, "property").flag());
     }
 
     @Override
-    public Set<RequestLogAvailability> availabilities() {
-        return RequestLogAvailabilitySet.of(flags);
+    public CompletableFuture<RequestLog> partialFuture(RequestLogProperty... properties) {
+        return future(RequestLogProperty.flags(requireNonNull(properties, "properties")));
     }
 
     @Override
-    public boolean isAvailable(RequestLogAvailability availability) {
-        return isAvailable(availability.getterFlags());
+    public CompletableFuture<RequestLog> partialFuture(Iterable<RequestLogProperty> properties) {
+        return future(RequestLogProperty.flags(requireNonNull(properties, "properties")));
     }
 
     @Override
-    public boolean isAvailable(RequestLogAvailability... availabilities) {
-        return isAvailable(getterFlags(availabilities));
+    public RequestLog ensureComplete() {
+        if (!isComplete()) {
+            throw new RequestLogAvailabilityException(RequestLog.class.getSimpleName() + " not complete");
+        }
+        return notCheckingAccessor;
     }
 
     @Override
-    public boolean isAvailable(Iterable<RequestLogAvailability> availabilities) {
-        return isAvailable(getterFlags(availabilities));
+    public RequestOnlyLog ensureRequestComplete() {
+        if (!isRequestComplete()) {
+            throw new RequestLogAvailabilityException(RequestOnlyLog.class.getSimpleName() + " not complete");
+        }
+        return notCheckingAccessor;
     }
 
-    private boolean isAvailable(int interestedFlags) {
-        return (flags & interestedFlags) == interestedFlags;
+    @Override
+    public RequestLog ensurePartial(RequestLogProperty property) {
+        if (!isAvailable(property)) {
+            throw new RequestLogAvailabilityException(property.name());
+        }
+        return partial();
```

Comment:
`return this` is probably fine for these partial methods, they'll usually access just a property or two anyways.